pandas/core/arrays/numpy_.py

```diff
@@ -91,15 +107,23 @@ def construct_array_type(cls):
         return PandasArray
 
     @property
-    def kind(self):
+    def kind(self) -> str:
+        """
+        A character code (one of 'biufcmMOSUV') identifying the general kind of data.
+        """
         return self._dtype.kind
 
     @property
-    def itemsize(self):
-        """The element size of this data-type object."""
+    def itemsize(self) -> int:
+        """
+        The element size of this data-type object.
+        """
         return self._dtype.itemsize
 
 
+PandasArrayT = TypeVar("PandasArrayT", bound="PandasArray")
```

Comment:
Why is this needed?

Reply:
L166 is type self, so needed for constructor signature to ensure values is type self.

Reply:
Don't we use the name as a string in other places?

Reply:
removed for now.

Reply:
To be clear, I didn't want to imply you necessarily needed to remove it, rather I want to understand it (eg what is the difference with using the class name as a string as we do elsewhere)

To me, this typing is adding a lot of black magic to the codebase, and I want to ensure it stays as understandable as possible for people (including me) that are not so familiar with typing.

Reply:
an example for the case where the return type is from a base class method

```
class Base:
    def foo(self) -> "Base":
        return type(self)()


class Derived(Base):
    x = 0

    def bar(self) -> int:
        return self.foo().x


print(Derived().bar())
```

This is valid python...

```
$ python pandas/test.py
0
```
however, mypy will report errors...
```
$ mypy pandas/test.py
pandas\test.py:10: error: "Base" has no attribute "x"
Found 1 error in 1 file (checked 1 source file)
```

whereas using a typevar...
```python
from typing import TypeVar

_T = TypeVar("_T")


class Base:
    def foo(self: _T) -> _T:
        return type(self)()


class Derived(Base):
    x = 0

    def bar(self) -> int:
        return self.foo().x


print(Derived().bar())
```
gives
```
$ mypy pandas/test.py
Success: no issues found in 1 source file
```

so typing the base class needs to use typevars to keep mypy green when we add types to the subclasses. I have reverted, but will probably need to add back when the subclasses get merged.

I've not included an example for the init case, but similar principal

Reply:
OK, that explanation helps (but if we start adding it, it needs to be explained in our contributing guidelines)

But: `PandasArray` is not a base class, it is not further subclassed I think. So why is it needed here?

Reply:
Ah, actually, StringArray now subclasses PandasArray, so forget the last sentence of what I just said ;)